File: core/ucp/apps/ucp-api/src/application/use-cases/toggle-user-status.use-case.ts
import { BadRequestException, Inject, Injectable, NotFoundException } from '@nestjs/common';
import type { ITenantRepository } from '../../ports/outbound/tenant.repository.js';
import { TENANT_REPOSITORY } from '../../ports/outbound/tenant.repository.js';
import type { IUserRepository } from '../../ports/outbound/user.repository.js';
import { USER_REPOSITORY } from '../../ports/outbound/user.repository.js';
import type { IUserIdentityProvider } from '../../ports/outbound/user-identity.provider.js';
import { USER_IDENTITY_PROVIDER } from '../../ports/outbound/user-identity.provider.js';

@Injectable()
export class ToggleUserStatusUseCase {
  constructor(
    @Inject(USER_REPOSITORY) private readonly userRepo: IUserRepository,
    @Inject(TENANT_REPOSITORY) private readonly tenantRepo: ITenantRepository,
    @Inject(USER_IDENTITY_PROVIDER) private readonly userIdentityProvider: IUserIdentityProvider,
  ) {}

  async execute(
    tenantId: string,
    userId: string,
    action: 'activate' | 'deactivate',
  ): Promise<void> {
    const tenant = await this.tenantRepo.findById(tenantId);
    if (!tenant) throw new NotFoundException('Tenant not found');

    if (!tenant.idpTenantId) {
      throw new BadRequestException('Tenant has no associated organization');
    }

    // Verify user belongs to this tenant
    const tenantUsers = await this.userRepo.findUsersByTenant(tenantId);
    const tenantUser = tenantUsers.find((u: { id: string }) => u.id === userId);
    if (!tenantUser || !tenantUser.idpUserId) {
      throw new NotFoundException('User identity mapping not found in this tenant');
    }

    // Load rich aggregate root
    const user = await this.userRepo.findById(userId);
    if (!user || !user.idpUserId) {
      throw new NotFoundException('User domain entity not found');
    }

    // 1. Sync external state first (Outbound Port)
    await this.userIdentityProvider.toggleUserStatus(user.idpUserId, tenant.idpTenantId, action);

    // 2. Perform business logic on the rich Domain Model
    if (action === 'activate') {
      user.activate();
    } else {
      user.deactivate();
    }

    // 3. Persist local state
    await this.userRepo.save(user);
  }
}

File: core/ucp/packages/ucp/src/ucp/application/use_cases/toggle_user_status_use_case.py
from dataclasses import dataclass
from typing import Literal

from ucp.domain.exceptions import ResourceNotFoundError
from ucp.ports.outbound.uow import UcpUnitOfWorkPort


@dataclass(frozen=True)
class ToggleUserStatusCommand:
    tenant_id: str
    user_id: str
    action: Literal["activate", "deactivate"]


class ToggleUserStatusUseCase:
    def __init__(
        self,
        uow: UcpUnitOfWorkPort,
    ):
        self._uow = uow

    async def execute(self, command: ToggleUserStatusCommand) -> None:
        async with self._uow:
            tenant = await self._uow.tenant_repo.find_by_id(command.tenant_id)
            if not tenant or not tenant.idp_tenant_id:
                raise ResourceNotFoundError(
                    f"Tenant {command.tenant_id} not found or missing IDP organization"
                )

            user = await self._uow.user_repo.find_by_id_and_tenant(
                user_id=command.user_id, tenant_id=command.tenant_id
            )

            if not user or not user.idp_user_id:
                raise ResourceNotFoundError(f"User mapping not found for {command.user_id}")

            # 1. Update local domain object state & Register Outbox Event
            user.change_status(action=command.action, tenant_id=command.tenant_id)
            await self._uow.user_repo.save(user)

            await self._uow.commit()
